File: priority-queue/typescript/priorityqueue.ts
export type Comparator<T> = (t1: T, t2: T) => -1 | 0 | 1

const DEFAULT_COMPARATOR: Comparator<any> = (a, b) => {
  return a < b ? -1 : a === b ? 0 : 1
}

export class PriorityQueue<T> {
  private items: T[]

  constructor(private readonly comparator: Comparator<T> = DEFAULT_COMPARATOR) {}

  isEmpty(): boolean {
    return this.items.length === 0
  }

  push(item: T): void {
    this.items.push(item)
    for (let i = this.items.length - 1; i > 0;) {
      const parent = Math.floor((i - 1) / 2)
      if (this.comparator(this.items[i], this.items[parent]) < 0) {
        this.swap(i, parent)
        i = parent
      } else {
        break
      }
    }
  }

  pop(): T | null {
    if (this.isEmpty()) {
      return null
    }

    const item = this.items[0]
    this.items[0] = this.items.pop()
    if (!this.isEmpty()) {
      this.heapify(0)
    }

    return item
  }

  top(): T | null {
    if (this.isEmpty()) {
      return null
    }

    return this.items[0]
  }

  private heapify(i: number): void {
    const l = 2 * i + 1
    const r = 2 * i + 2
    let smallest = i
    const size = this.items.length
    if (l < size && this.comparator(this.items[l], this.items[i]) < 0) {
      smallest = l
    }
    if (r < size && this.comparator(this.items[r], this.items[smallest]) < 0) {
      smallest = r
    }
    if (smallest !== i) {
      this.swap(i, smallest)
      this.heapify(smallest)
    }
  }

  private swap(i: number, j: number): void {
    const tmp = this.items[i]
    this.items[i] = this.items[j]
    this.items[j] = tmp
  }
}

File: src/go/priorityqueue/priorityqueue.go
package priorityqueue

// Item represents a single object in the heap.
type Item interface {
	// Less tests whether the current item is less than the given argument.
	Less(i Item) bool
}

// Heap structure
type Heap struct {
	items []Item
}

// New creates empty heap
func New() *Heap {
	return &Heap{}
}

// IsEmpty determines if heap empty or not
func (h *Heap) IsEmpty() bool {
	return len(h.items) == 0
}

// Push pushes 1 item to the heap and rebuild
func (h *Heap) Push(item Item) {
	h.items = append(h.items, item)

	i := len(h.items) - 1
	for i > 0 {
		parent := (i - 1) / 2
		if h.items[i].Less(h.items[parent]) {
			swap(&h.items[i], &h.items[parent])
			i = parent
		} else {
			break
		}
	}
}

// Pop removes the first item and return it
func (h *Heap) Pop() Item {
	if h.IsEmpty() {
		return nil
	}

	item := h.items[0]
	index := len(h.items) - 1
	h.items[0] = h.items[index]
	h.items[index] = nil
	h.items = h.items[:index]
	if !h.IsEmpty() {
		h.heapify(0)
	}

	return item
}

// Top gets the first item without remove it
func (h *Heap) Top() Item {
	if h.IsEmpty() {
		return nil
	}

	return h.items[0]
}

func (h *Heap) heapify(i int) {
	l := 2*i + 1
	r := 2*i + 2
	smallest := i
	heapSize := len(h.items)
	if l < heapSize && h.items[l].Less(h.items[i]) {
		smallest = l
	}
	if r < heapSize && h.items[r].Less(h.items[smallest]) {
		smallest = r
	}
	if smallest != i {
		swap(&h.items[i], &h.items[smallest])
		h.heapify(smallest)
	}
}

func swap(i1 *Item, i2 *Item) {
	tmp := *i1
	*i1 = *i2
	*i2 = tmp
}
